test/integration/scheduler_perf/scheduler_perf_test.go

```diff
@@ -768,84 +862,155 @@ func unrollWorkloadTemplate(b *testing.B, wt []op, w *workload) []op {
+	b, benchmarking := tb.(*testing.B)
+	if benchmarking {
+		start := time.Now()
+		b.Cleanup(func() {
+			duration := time.Since(start)
+			// This includes startup and shutdown time and thus does not
+			// reflect scheduling performance. It's useful to get a feeling
+			// for how long each workload runs overall.
+			b.ReportMetric(duration.Seconds(), "runtime_seconds")
+		})
+	}
+
+	// Disable error checking of the sampling interval length in the
+	// throughput collector by default. When running benchmarks, report
+	// it as test failure when samples are not taken regularly.
+	var throughputErrorMargin float64
+	if benchmarking {
+		// TODO: To prevent the perf-test failure, we increased the error margin, if still not enough
+		// one day, we should think of another approach to avoid this trick.
+		throughputErrorMargin = 30
+	}
 
 	// Additional informers needed for testing. The pod informer was
 	// already created before (scheduler.NewInformerFactory) and the
 	// factory was started for it (mustSetupScheduler), therefore we don't
 	// need to start again.
 	podInformer := informerFactory.Core().V1().Pods()
 
+	// Everything else started by this function gets stopped before it returns.
+	ctx, cancel := context.WithCancel(ctx)
+	var wg sync.WaitGroup
+	defer wg.Wait()
+	defer cancel()
+
 	var mu sync.Mutex
 	var dataItems []DataItem
 	nextNodeIndex := 0
 	// numPodsScheduledPerNamespace has all namespaces created in workload and the number of pods they (will) have.
 	// All namespaces listed in numPodsScheduledPerNamespace will be cleaned up.
 	numPodsScheduledPerNamespace := make(map[string]int)
 
-	b.Cleanup(func() {
-		for namespace := range numPodsScheduledPerNamespace {
-			if err := client.CoreV1().Namespaces().Delete(ctx, namespace, metav1.DeleteOptions{}); err != nil {
-				b.Errorf("Deleting Namespace in numPodsScheduledPerNamespace: %v", err)
+	if cleanup {
+		// This must run before controllers get cleaned up.
+		defer func() {
+			deleteNow := *metav1.NewDeleteOptions(0)
+			for namespace := range numPodsScheduledPerNamespace {
+				// Pods have to be deleted explicitly, with no grace period. Normally
+				// kubelet will set the DeletionGracePeriodSeconds to zero when it's okay
+				// to remove a deleted pod, but we don't run kubelet...
+				pods, err := client.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions{})
+				if err != nil {
+					tb.Fatalf("failed to list pods in %q: %v", namespace, err)
+				}
+				for _, pod := range pods.Items {
+					if err := client.CoreV1().Pods(namespace).Delete(ctx, pod.Name, deleteNow); err != nil {
+						tb.Fatalf("failed to delete pod %q in namespace %q: %v", pod.Name, namespace, err)
+					}
+				}
+				if err := client.CoreV1().Namespaces().Delete(ctx, namespace, deleteNow); err != nil {
+					tb.Fatalf("Deleting Namespace in numPodsScheduledPerNamespace: %v", err)
+				}
 			}
-		}
-	})
+			gomega.NewGomegaWithT(tb).Eventually(ctx, func(ctx context.Context) ([]interface{}, error) {
```

Comment:
I don't see why we need to compare `objects` here.

This is a cleanup function. The pair `bool, error` from `wait` seems enough.

Reply:
I added some more inline comments about the motivation for this code.